api/v1_proof.inc

```diff
@@ -0,0 +1,388 @@
+<?php
+include_once($relPath.'Project.inc'); // can_be_proofed_by_current_user()
+include_once($relPath.'stages.inc'); // can_user_get_pages_in_project()
+include_once($relPath.'DPage.inc'); // _log_page_event(), _project_adjust_n_available_pages()
+include_once($relPath.'RoundDescriptor.inc'); // get_Round_for_project_state(), is_a_mentor_round()
+include_once("exceptions.inc");
+
+// checkout a page to a user
+// update the project table in one operation, then select the page we have
+// checked out. (This could be modified to enable checking out several pages)
+// Copying the text from previous round is different to previous action.
+// POST because not idempotent
+// POST v1/projects/:projectid/projectstates/:projectstateid/checkout
+// (similar to LPage get_available_page())
+function api_v1_project_state_checkout($method, $data, $query_params)
+{
+    global $pguser, $preceding_proofer_restriction;
```

Comment:
We need to figure out how to reuse existing LPage functionality or do some refactoring rather than duplicating logic. This is a bit of a problem already in the API code (most of it my fault) and I don't want to do that for very critical code that manipulates data.

If we were going to do a wholesale cutover and remove the old way and only use the new way then I wouldn't worry about it, but I think doing a successful UI redesign is going to be a staged process that involves rolling out the new API alongside the current UI, then developing the new UI to use the API, then removing the old UI.

Part of the heavy lift on Task 2011 is that I'm pulling out the project validation logic into a single place for re-use in the UI and the API.

Reply:
I'll have a look into LPage.